{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> This notebook contains the evolutionary dataset first analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import re\n",
    "import csv\n",
    "\n",
    "from Bio import AlignIO\n",
    "\n",
    "from Bio.Align import AlignInfo\n",
    "\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "import os\n",
    "from helpers import*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Data Loading"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> We have both xls and fasta format for the evolutionary dataset. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data Paths\n",
    "FASTA_msa_seq = 'original_data/msa/4.Fasta_all_7k.txt'\n",
    "XLS_msa_seq = 'original_data/msa/Dataset_S4.xls'\n",
    "XLS_msa_seq_csv = 'original_data/msa/Dataset_S4.csv'\n",
    "\n",
    "# Read the Excel file and write to CSV\n",
    "df = pd.read_excel(XLS_msa_seq)\n",
    "df.to_csv(XLS_msa_seq_csv, index=False)\n",
    "XLS_msa = pd.read_csv(XLS_msa_seq_csv)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The minimum value in the second column is: -3.743276064\n",
      "The maximum value in the second column is: 0.3557906625\n",
      "The number of lines in the CSV file is: 4807\n"
     ]
    }
   ],
   "source": [
    "# Loading dataframe\n",
    "file_path = 'data/BLAT_ECOLX_Ranganathan2015-2500/data.csv'\n",
    "df = pd.read_csv(file_path)\n",
    "\n",
    "second_column = df.iloc[:, 1]\n",
    "\n",
    "# Find the minimum and maximum values\n",
    "min_value = second_column.min()\n",
    "max_value = second_column.max()\n",
    "number_of_lines = len(df.index)\n",
    "\n",
    "print(f\"The minimum value in the second column is: {min_value}\")\n",
    "print(f\"The maximum value in the second column is: {max_value}\")\n",
    "print(f\"The number of lines in the CSV file is: {number_of_lines}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of '>' instances in the file is: 9255\n"
     ]
    }
   ],
   "source": [
    "file_path1 = 'alignments/Q2NB98.a2m'\n",
    "\n",
    "\n",
    "count = 0\n",
    "\n",
    "# Open the file and count '>' instances\n",
    "with open(file_path1, 'r') as file:\n",
    "    for line in file:\n",
    "        # Increment count if '>' is found at the beginning of a line\n",
    "        if line.startswith('>'):\n",
    "            count += 1\n",
    "\n",
    "print(f\"The number of '>' instances in the file is: {count}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>Unnamed: 2</th>\n",
       "      <th>Unnamed: 3</th>\n",
       "      <th>Unnamed: 4</th>\n",
       "      <th>Unnamed: 5</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "      <th>Unnamed: 7</th>\n",
       "      <th>Unnamed: 8</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>Unnamed: 11</th>\n",
       "      <th>Unnamed: 12</th>\n",
       "      <th>Unnamed: 13</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>Unnamed: 15</th>\n",
       "      <th>Unnamed: 16</th>\n",
       "      <th>Unnamed: 17</th>\n",
       "      <th>Unnamed: 18</th>\n",
       "      <th>Unnamed: 19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>User note: GenBank Accession numbers marked wi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Database Source</td>\n",
       "      <td>Sequence ID</td>\n",
       "      <td>GenBank Accession</td>\n",
       "      <td>Primary Structure</td>\n",
       "      <td>GXNCRFLQ Motif</td>\n",
       "      <td>Protein Length</td>\n",
       "      <td>Domain Structure</td>\n",
       "      <td>Functional Cluster</td>\n",
       "      <td>Primary Effector</td>\n",
       "      <td>Primary Effector Gene Ontology</td>\n",
       "      <td>Linker Length</td>\n",
       "      <td>Number of Predicted Transmembrane Helices</td>\n",
       "      <td>Transmembrane Helix Topology</td>\n",
       "      <td>Kingdom</td>\n",
       "      <td>Phylum</td>\n",
       "      <td>Class</td>\n",
       "      <td>Order</td>\n",
       "      <td>Family</td>\n",
       "      <td>Genus</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Interpro</td>\n",
       "      <td>A0A009EVU7</td>\n",
       "      <td>EXA65901.1</td>\n",
       "      <td>MTSFYNGGTSQEIFNLIGQAAALEDIIDAITVWLGDKIPDALVSVM...</td>\n",
       "      <td>GKNCRILQ</td>\n",
       "      <td>850.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; GAF_2 : (9, 166) ---...</td>\n",
       "      <td>['EAL', 'GAF', 'GGDEF', 'LOV', 'PAS']</td>\n",
       "      <td>GGDEF</td>\n",
       "      <td>Cyclic Nucleotide Biosythesis</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Moraxellaceae</td>\n",
       "      <td>Acinetobacter</td>\n",
       "      <td>baumannii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Interpro</td>\n",
       "      <td>A0A009JQP1</td>\n",
       "      <td>EXB31618.1</td>\n",
       "      <td>MNNNCDDHDTYHIFNLIGQHTELHLILSAASTWLESRISNVLVAIM...</td>\n",
       "      <td>GKTCHFLQ</td>\n",
       "      <td>855.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; GAF_2 : (26, 166) --...</td>\n",
       "      <td>['EAL', 'GAF', 'GGDEF', 'LOV', 'PAS']</td>\n",
       "      <td>GGDEF</td>\n",
       "      <td>Cyclic Nucleotide Biosythesis</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Moraxellaceae</td>\n",
       "      <td>Acinetobacter</td>\n",
       "      <td>baumannii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6780</th>\n",
       "      <td>OneKP</td>\n",
       "      <td>ZYAX_2075737</td>\n",
       "      <td>KU702188</td>\n",
       "      <td>MEPSNPGNTNNSFPSGNVVRKVLDDSPAIISPLPRDSRGSLEVFNP...</td>\n",
       "      <td>GKNCRFLQ</td>\n",
       "      <td>1083.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; TLOV : (286, 663) --...</td>\n",
       "      <td>['Pkinase', 'TLOV']</td>\n",
       "      <td>Pkinase</td>\n",
       "      <td>Kinase Activity</td>\n",
       "      <td>83.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Land Plants</td>\n",
       "      <td>Pinophyta</td>\n",
       "      <td>Pinopsida</td>\n",
       "      <td>Pinales</td>\n",
       "      <td>Taxaceae</td>\n",
       "      <td>Taxus</td>\n",
       "      <td>cuspidata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6781</th>\n",
       "      <td>OneKP</td>\n",
       "      <td>ZZOL_2003873</td>\n",
       "      <td>KU702189</td>\n",
       "      <td>MESTSGSKQDHYFSQRSSSMPKPVRDSLAQLPRDSRGSLEVFNPQG...</td>\n",
       "      <td>GRNCRFLQ</td>\n",
       "      <td>1008.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; TLOV : (232, 597) --...</td>\n",
       "      <td>['Pkinase', 'TLOV']</td>\n",
       "      <td>Pkinase</td>\n",
       "      <td>Kinase Activity</td>\n",
       "      <td>83.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Land Plants</td>\n",
       "      <td>Lycopodiophyta</td>\n",
       "      <td>Isoetopsida</td>\n",
       "      <td>Selaginellales</td>\n",
       "      <td>Selaginellaceae</td>\n",
       "      <td>Selaginella</td>\n",
       "      <td>lepidophylla</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6782</th>\n",
       "      <td>OneKP</td>\n",
       "      <td>ZZOL_2007563</td>\n",
       "      <td>KU702190</td>\n",
       "      <td>MALQNGSLLSSDRKCVRSPRMKNHKLIVSQSLIRSYRASIQDELQK...</td>\n",
       "      <td>GRNCRFLQ</td>\n",
       "      <td>394.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; TLOV : (49, 354) ---...</td>\n",
       "      <td>['TLOV']</td>\n",
       "      <td>Short_LOV</td>\n",
       "      <td>Short LOV</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Land Plants</td>\n",
       "      <td>Lycopodiophyta</td>\n",
       "      <td>Isoetopsida</td>\n",
       "      <td>Selaginellales</td>\n",
       "      <td>Selaginellaceae</td>\n",
       "      <td>Selaginella</td>\n",
       "      <td>lepidophylla</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6783</th>\n",
       "      <td>OneKP</td>\n",
       "      <td>ZZOL_2011001</td>\n",
       "      <td>KU702191#</td>\n",
       "      <td>MLVEVSQFTEGSKDKAVRPNGMPESLIKYDARQKDSVVSSVSELVE...</td>\n",
       "      <td>GKNCRFLQ</td>\n",
       "      <td>465.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; LOV : (142, 247) ---...</td>\n",
       "      <td>['LOV', 'Pkinase']</td>\n",
       "      <td>Pkinase</td>\n",
       "      <td>Kinase Activity</td>\n",
       "      <td>83.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Land Plants</td>\n",
       "      <td>Lycopodiophyta</td>\n",
       "      <td>Isoetopsida</td>\n",
       "      <td>Selaginellales</td>\n",
       "      <td>Selaginellaceae</td>\n",
       "      <td>Selaginella</td>\n",
       "      <td>lepidophylla</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6784</th>\n",
       "      <td>OneKP</td>\n",
       "      <td>ZZOL_2011476</td>\n",
       "      <td>KU702192</td>\n",
       "      <td>MDVDVELDSDNSSSLGSYGDADAGFQFEAHNTTAPNLSFFGFDASV...</td>\n",
       "      <td>GRNCRFLQ</td>\n",
       "      <td>638.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; LOV : (55, 166) ---&gt;...</td>\n",
       "      <td>['F-box', 'Kelch', 'LOV']</td>\n",
       "      <td>F-box</td>\n",
       "      <td>Protein interaction</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Land Plants</td>\n",
       "      <td>Lycopodiophyta</td>\n",
       "      <td>Isoetopsida</td>\n",
       "      <td>Selaginellales</td>\n",
       "      <td>Selaginellaceae</td>\n",
       "      <td>Selaginella</td>\n",
       "      <td>lepidophylla</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6785 rows Ã— 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Unnamed: 0                                         Unnamed: 1  \\\n",
       "0                 NaN  User note: GenBank Accession numbers marked wi...   \n",
       "1                 NaN                                                NaN   \n",
       "2     Database Source                                        Sequence ID   \n",
       "3            Interpro                                         A0A009EVU7   \n",
       "4            Interpro                                         A0A009JQP1   \n",
       "...               ...                                                ...   \n",
       "6780            OneKP                                       ZYAX_2075737   \n",
       "6781            OneKP                                       ZZOL_2003873   \n",
       "6782            OneKP                                       ZZOL_2007563   \n",
       "6783            OneKP                                       ZZOL_2011001   \n",
       "6784            OneKP                                       ZZOL_2011476   \n",
       "\n",
       "             Unnamed: 2                                         Unnamed: 3  \\\n",
       "0                   NaN                                                NaN   \n",
       "1                   NaN                                                NaN   \n",
       "2     GenBank Accession                                  Primary Structure   \n",
       "3            EXA65901.1  MTSFYNGGTSQEIFNLIGQAAALEDIIDAITVWLGDKIPDALVSVM...   \n",
       "4            EXB31618.1  MNNNCDDHDTYHIFNLIGQHTELHLILSAASTWLESRISNVLVAIM...   \n",
       "...                 ...                                                ...   \n",
       "6780           KU702188  MEPSNPGNTNNSFPSGNVVRKVLDDSPAIISPLPRDSRGSLEVFNP...   \n",
       "6781           KU702189  MESTSGSKQDHYFSQRSSSMPKPVRDSLAQLPRDSRGSLEVFNPQG...   \n",
       "6782           KU702190  MALQNGSLLSSDRKCVRSPRMKNHKLIVSQSLIRSYRASIQDELQK...   \n",
       "6783          KU702191#  MLVEVSQFTEGSKDKAVRPNGMPESLIKYDARQKDSVVSSVSELVE...   \n",
       "6784           KU702192  MDVDVELDSDNSSSLGSYGDADAGFQFEAHNTTAPNLSFFGFDASV...   \n",
       "\n",
       "          Unnamed: 4      Unnamed: 5  \\\n",
       "0                NaN             NaN   \n",
       "1                NaN             NaN   \n",
       "2     GXNCRFLQ Motif  Protein Length   \n",
       "3           GKNCRILQ           850.0   \n",
       "4           GKTCHFLQ           855.0   \n",
       "...              ...             ...   \n",
       "6780        GKNCRFLQ          1083.0   \n",
       "6781        GRNCRFLQ          1008.0   \n",
       "6782        GRNCRFLQ           394.0   \n",
       "6783        GKNCRFLQ           465.0   \n",
       "6784        GRNCRFLQ           638.0   \n",
       "\n",
       "                                             Unnamed: 6  \\\n",
       "0                                                   NaN   \n",
       "1                                                   NaN   \n",
       "2                                      Domain Structure   \n",
       "3     [N_terminus : (0, 0) ---> GAF_2 : (9, 166) ---...   \n",
       "4     [N_terminus : (0, 0) ---> GAF_2 : (26, 166) --...   \n",
       "...                                                 ...   \n",
       "6780  [N_terminus : (0, 0) ---> TLOV : (286, 663) --...   \n",
       "6781  [N_terminus : (0, 0) ---> TLOV : (232, 597) --...   \n",
       "6782  [N_terminus : (0, 0) ---> TLOV : (49, 354) ---...   \n",
       "6783  [N_terminus : (0, 0) ---> LOV : (142, 247) ---...   \n",
       "6784  [N_terminus : (0, 0) ---> LOV : (55, 166) --->...   \n",
       "\n",
       "                                 Unnamed: 7        Unnamed: 8  \\\n",
       "0                                       NaN               NaN   \n",
       "1                                       NaN               NaN   \n",
       "2                        Functional Cluster  Primary Effector   \n",
       "3     ['EAL', 'GAF', 'GGDEF', 'LOV', 'PAS']             GGDEF   \n",
       "4     ['EAL', 'GAF', 'GGDEF', 'LOV', 'PAS']             GGDEF   \n",
       "...                                     ...               ...   \n",
       "6780                    ['Pkinase', 'TLOV']           Pkinase   \n",
       "6781                    ['Pkinase', 'TLOV']           Pkinase   \n",
       "6782                               ['TLOV']         Short_LOV   \n",
       "6783                     ['LOV', 'Pkinase']           Pkinase   \n",
       "6784              ['F-box', 'Kelch', 'LOV']             F-box   \n",
       "\n",
       "                          Unnamed: 9    Unnamed: 10  \\\n",
       "0                                NaN            NaN   \n",
       "1                                NaN            NaN   \n",
       "2     Primary Effector Gene Ontology  Linker Length   \n",
       "3      Cyclic Nucleotide Biosythesis           12.0   \n",
       "4      Cyclic Nucleotide Biosythesis           12.0   \n",
       "...                              ...            ...   \n",
       "6780                 Kinase Activity           83.0   \n",
       "6781                 Kinase Activity           83.0   \n",
       "6782                       Short LOV            0.0   \n",
       "6783                 Kinase Activity           83.0   \n",
       "6784             Protein interaction           56.0   \n",
       "\n",
       "                                    Unnamed: 11  \\\n",
       "0                                           NaN   \n",
       "1                                           NaN   \n",
       "2     Number of Predicted Transmembrane Helices   \n",
       "3                                           0.0   \n",
       "4                                           0.0   \n",
       "...                                         ...   \n",
       "6780                                        0.0   \n",
       "6781                                        0.0   \n",
       "6782                                        0.0   \n",
       "6783                                        0.0   \n",
       "6784                                        0.0   \n",
       "\n",
       "                        Unnamed: 12  Unnamed: 13     Unnamed: 14  \\\n",
       "0                               NaN          NaN             NaN   \n",
       "1                               NaN          NaN             NaN   \n",
       "2     Transmembrane Helix Topology       Kingdom          Phylum   \n",
       "3                               NaN     Bacteria  Proteobacteria   \n",
       "4                               NaN     Bacteria  Proteobacteria   \n",
       "...                             ...          ...             ...   \n",
       "6780                            NaN  Land Plants       Pinophyta   \n",
       "6781                            NaN  Land Plants  Lycopodiophyta   \n",
       "6782                            NaN  Land Plants  Lycopodiophyta   \n",
       "6783                            NaN  Land Plants  Lycopodiophyta   \n",
       "6784                            NaN  Land Plants  Lycopodiophyta   \n",
       "\n",
       "              Unnamed: 15      Unnamed: 16      Unnamed: 17    Unnamed: 18  \\\n",
       "0                     NaN              NaN              NaN            NaN   \n",
       "1                     NaN              NaN              NaN            NaN   \n",
       "2                   Class            Order           Family          Genus   \n",
       "3     Gammaproteobacteria  Pseudomonadales    Moraxellaceae  Acinetobacter   \n",
       "4     Gammaproteobacteria  Pseudomonadales    Moraxellaceae  Acinetobacter   \n",
       "...                   ...              ...              ...            ...   \n",
       "6780            Pinopsida          Pinales         Taxaceae          Taxus   \n",
       "6781          Isoetopsida   Selaginellales  Selaginellaceae    Selaginella   \n",
       "6782          Isoetopsida   Selaginellales  Selaginellaceae    Selaginella   \n",
       "6783          Isoetopsida   Selaginellales  Selaginellaceae    Selaginella   \n",
       "6784          Isoetopsida   Selaginellales  Selaginellaceae    Selaginella   \n",
       "\n",
       "       Unnamed: 19  \n",
       "0              NaN  \n",
       "1              NaN  \n",
       "2          Species  \n",
       "3        baumannii  \n",
       "4        baumannii  \n",
       "...            ...  \n",
       "6780     cuspidata  \n",
       "6781  lepidophylla  \n",
       "6782  lepidophylla  \n",
       "6783  lepidophylla  \n",
       "6784  lepidophylla  \n",
       "\n",
       "[6785 rows x 20 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(XLS_msa)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  2. Convert MSA .a2m file to adequate .a2m format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> The alignment requirements are the folowing:\n",
    "The first step of the EVmutation method is to compute a pairwise model of sequences from a family sequence alignment, e.g. the included example file (example/msa_comparison.a2m). This alignment has to be in aligned FASTA/A2M format and must fulfill the following requirements:\n",
    "* The target sequence may not contain any gaps.\n",
    "* Columns that should be excluded from model inference (e.g. too many gaps) must be represented by lowercase characters. Gaps in these positions must be represented by a dot (\".\") rather than a dash (\"-\").\n",
    "* The identifier of the target sequence has to be passed to plmc with the -f parameter (see below)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> The initial .a2m format is generated using the hh-suite repo (which should be downloaded): https://github.com/soedinglab/hh-suite, which allows for conversion between different MSA file formats. The command that should be used is the following where the reformat function is uised: reformat.pl file_to_be_converted.a3m converted_file.a2m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "initial_a2m_format = 'alignments/LOV_Q2NB98.a2m'\n",
    "correct_a2m_format = 'alignments/Q2NB98.a2m'\n",
    "convert_to_correct_a2m(initial_a2m_format, correct_a2m_format)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ">Let's apply this function to the fasta file (LOV domains starting with EL222 aligned.fasta)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence Homology: 0.0%\n"
     ]
    }
   ],
   "source": [
    "homology = calculate_sequence_homology('alignments/Q2NB98.a2m')\n",
    "print(f\"Sequence Homology: {homology}%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sequences removed: 4363\n",
      "Percentage of sequences removed: 47.14%\n"
     ]
    }
   ],
   "source": [
    "LOV_msa_a2m_30gaps_removed = 'alignments/Q2NB98_30perc.a2m'\n",
    "LOV_msa_a2m_dashes = 'alignments/Q2NB98.a2m'\n",
    "num_removed, perc_removed = remove_high_gap_sequences(LOV_msa_a2m_dashes, LOV_msa_a2m_30gaps_removed, gap_threshold=30)\n",
    "\n",
    "print(f\"Number of sequences removed: {num_removed}\")\n",
    "print(f\"Percentage of sequences removed: {perc_removed:.2f}%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  3. Domain of interest range analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lov_data = extract_lov_domain_tuples(XLS_msa)\n",
    "# print(lov_data)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ada",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
