{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> This notebook contains the evolutionary dataset first analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Data Loading"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> We have both xls and fasta format for the evolutionary dataset. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data Paths\n",
    "FASTA_msa_seq = 'original_data/msa/4.Fasta_all_7k.txt'\n",
    "XLS_msa_seq = 'original_data/msa/Dataset_S4.xls'\n",
    "XLS_msa_seq_csv = 'original_data/msa/Dataset_S4.csv'\n",
    "\n",
    "# Read the Excel file and write to CSV\n",
    "df = pd.read_excel(XLS_msa_seq)\n",
    "df.to_csv(XLS_msa_seq_csv, index=False)\n",
    "\n",
    "# Data Loading\n",
    "#fasta_msa = pd.read_csv(FASTA_msa_seq,  delimiter='\\t')\n",
    "XLS_msa = pd.read_csv(XLS_msa_seq_csv)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>Unnamed: 2</th>\n",
       "      <th>Unnamed: 3</th>\n",
       "      <th>Unnamed: 4</th>\n",
       "      <th>Unnamed: 5</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "      <th>Unnamed: 7</th>\n",
       "      <th>Unnamed: 8</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>Unnamed: 11</th>\n",
       "      <th>Unnamed: 12</th>\n",
       "      <th>Unnamed: 13</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>Unnamed: 15</th>\n",
       "      <th>Unnamed: 16</th>\n",
       "      <th>Unnamed: 17</th>\n",
       "      <th>Unnamed: 18</th>\n",
       "      <th>Unnamed: 19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>User note: GenBank Accession numbers marked wi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Database Source</td>\n",
       "      <td>Sequence ID</td>\n",
       "      <td>GenBank Accession</td>\n",
       "      <td>Primary Structure</td>\n",
       "      <td>GXNCRFLQ Motif</td>\n",
       "      <td>Protein Length</td>\n",
       "      <td>Domain Structure</td>\n",
       "      <td>Functional Cluster</td>\n",
       "      <td>Primary Effector</td>\n",
       "      <td>Primary Effector Gene Ontology</td>\n",
       "      <td>Linker Length</td>\n",
       "      <td>Number of Predicted Transmembrane Helices</td>\n",
       "      <td>Transmembrane Helix Topology</td>\n",
       "      <td>Kingdom</td>\n",
       "      <td>Phylum</td>\n",
       "      <td>Class</td>\n",
       "      <td>Order</td>\n",
       "      <td>Family</td>\n",
       "      <td>Genus</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Interpro</td>\n",
       "      <td>A0A009EVU7</td>\n",
       "      <td>EXA65901.1</td>\n",
       "      <td>MTSFYNGGTSQEIFNLIGQAAALEDIIDAITVWLGDKIPDALVSVM...</td>\n",
       "      <td>GKNCRILQ</td>\n",
       "      <td>850.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; GAF_2 : (9, 166) ---...</td>\n",
       "      <td>['EAL', 'GAF', 'GGDEF', 'LOV', 'PAS']</td>\n",
       "      <td>GGDEF</td>\n",
       "      <td>Cyclic Nucleotide Biosythesis</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Moraxellaceae</td>\n",
       "      <td>Acinetobacter</td>\n",
       "      <td>baumannii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Interpro</td>\n",
       "      <td>A0A009JQP1</td>\n",
       "      <td>EXB31618.1</td>\n",
       "      <td>MNNNCDDHDTYHIFNLIGQHTELHLILSAASTWLESRISNVLVAIM...</td>\n",
       "      <td>GKTCHFLQ</td>\n",
       "      <td>855.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; GAF_2 : (26, 166) --...</td>\n",
       "      <td>['EAL', 'GAF', 'GGDEF', 'LOV', 'PAS']</td>\n",
       "      <td>GGDEF</td>\n",
       "      <td>Cyclic Nucleotide Biosythesis</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Moraxellaceae</td>\n",
       "      <td>Acinetobacter</td>\n",
       "      <td>baumannii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6780</th>\n",
       "      <td>OneKP</td>\n",
       "      <td>ZYAX_2075737</td>\n",
       "      <td>KU702188</td>\n",
       "      <td>MEPSNPGNTNNSFPSGNVVRKVLDDSPAIISPLPRDSRGSLEVFNP...</td>\n",
       "      <td>GKNCRFLQ</td>\n",
       "      <td>1083.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; TLOV : (286, 663) --...</td>\n",
       "      <td>['Pkinase', 'TLOV']</td>\n",
       "      <td>Pkinase</td>\n",
       "      <td>Kinase Activity</td>\n",
       "      <td>83.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>Land Plants</td>\n",
       "      <td>Pinophyta</td>\n",
       "      <td>Pinopsida</td>\n",
       "      <td>Pinales</td>\n",
       "      <td>Taxaceae</td>\n",
       "      <td>Taxus</td>\n",
       "      <td>cuspidata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6781</th>\n",
       "      <td>OneKP</td>\n",
       "      <td>ZZOL_2003873</td>\n",
       "      <td>KU702189</td>\n",
       "      <td>MESTSGSKQDHYFSQRSSSMPKPVRDSLAQLPRDSRGSLEVFNPQG...</td>\n",
       "      <td>GRNCRFLQ</td>\n",
       "      <td>1008.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; TLOV : (232, 597) --...</td>\n",
       "      <td>['Pkinase', 'TLOV']</td>\n",
       "      <td>Pkinase</td>\n",
       "      <td>Kinase Activity</td>\n",
       "      <td>83.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>Land Plants</td>\n",
       "      <td>Lycopodiophyta</td>\n",
       "      <td>Isoetopsida</td>\n",
       "      <td>Selaginellales</td>\n",
       "      <td>Selaginellaceae</td>\n",
       "      <td>Selaginella</td>\n",
       "      <td>lepidophylla</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6782</th>\n",
       "      <td>OneKP</td>\n",
       "      <td>ZZOL_2007563</td>\n",
       "      <td>KU702190</td>\n",
       "      <td>MALQNGSLLSSDRKCVRSPRMKNHKLIVSQSLIRSYRASIQDELQK...</td>\n",
       "      <td>GRNCRFLQ</td>\n",
       "      <td>394.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; TLOV : (49, 354) ---...</td>\n",
       "      <td>['TLOV']</td>\n",
       "      <td>Short_LOV</td>\n",
       "      <td>Short LOV</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>Land Plants</td>\n",
       "      <td>Lycopodiophyta</td>\n",
       "      <td>Isoetopsida</td>\n",
       "      <td>Selaginellales</td>\n",
       "      <td>Selaginellaceae</td>\n",
       "      <td>Selaginella</td>\n",
       "      <td>lepidophylla</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6783</th>\n",
       "      <td>OneKP</td>\n",
       "      <td>ZZOL_2011001</td>\n",
       "      <td>KU702191#</td>\n",
       "      <td>MLVEVSQFTEGSKDKAVRPNGMPESLIKYDARQKDSVVSSVSELVE...</td>\n",
       "      <td>GKNCRFLQ</td>\n",
       "      <td>465.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; LOV : (142, 247) ---...</td>\n",
       "      <td>['LOV', 'Pkinase']</td>\n",
       "      <td>Pkinase</td>\n",
       "      <td>Kinase Activity</td>\n",
       "      <td>83.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>Land Plants</td>\n",
       "      <td>Lycopodiophyta</td>\n",
       "      <td>Isoetopsida</td>\n",
       "      <td>Selaginellales</td>\n",
       "      <td>Selaginellaceae</td>\n",
       "      <td>Selaginella</td>\n",
       "      <td>lepidophylla</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6784</th>\n",
       "      <td>OneKP</td>\n",
       "      <td>ZZOL_2011476</td>\n",
       "      <td>KU702192</td>\n",
       "      <td>MDVDVELDSDNSSSLGSYGDADAGFQFEAHNTTAPNLSFFGFDASV...</td>\n",
       "      <td>GRNCRFLQ</td>\n",
       "      <td>638.0</td>\n",
       "      <td>[N_terminus : (0, 0) ---&gt; LOV : (55, 166) ---&gt;...</td>\n",
       "      <td>['F-box', 'Kelch', 'LOV']</td>\n",
       "      <td>F-box</td>\n",
       "      <td>Protein interaction</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>Land Plants</td>\n",
       "      <td>Lycopodiophyta</td>\n",
       "      <td>Isoetopsida</td>\n",
       "      <td>Selaginellales</td>\n",
       "      <td>Selaginellaceae</td>\n",
       "      <td>Selaginella</td>\n",
       "      <td>lepidophylla</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6785 rows Ã— 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Unnamed: 0                                         Unnamed: 1  \\\n",
       "0                 NaN  User note: GenBank Accession numbers marked wi...   \n",
       "1                 NaN                                                NaN   \n",
       "2     Database Source                                        Sequence ID   \n",
       "3            Interpro                                         A0A009EVU7   \n",
       "4            Interpro                                         A0A009JQP1   \n",
       "...               ...                                                ...   \n",
       "6780            OneKP                                       ZYAX_2075737   \n",
       "6781            OneKP                                       ZZOL_2003873   \n",
       "6782            OneKP                                       ZZOL_2007563   \n",
       "6783            OneKP                                       ZZOL_2011001   \n",
       "6784            OneKP                                       ZZOL_2011476   \n",
       "\n",
       "             Unnamed: 2                                         Unnamed: 3  \\\n",
       "0                   NaN                                                NaN   \n",
       "1                   NaN                                                NaN   \n",
       "2     GenBank Accession                                  Primary Structure   \n",
       "3            EXA65901.1  MTSFYNGGTSQEIFNLIGQAAALEDIIDAITVWLGDKIPDALVSVM...   \n",
       "4            EXB31618.1  MNNNCDDHDTYHIFNLIGQHTELHLILSAASTWLESRISNVLVAIM...   \n",
       "...                 ...                                                ...   \n",
       "6780           KU702188  MEPSNPGNTNNSFPSGNVVRKVLDDSPAIISPLPRDSRGSLEVFNP...   \n",
       "6781           KU702189  MESTSGSKQDHYFSQRSSSMPKPVRDSLAQLPRDSRGSLEVFNPQG...   \n",
       "6782           KU702190  MALQNGSLLSSDRKCVRSPRMKNHKLIVSQSLIRSYRASIQDELQK...   \n",
       "6783          KU702191#  MLVEVSQFTEGSKDKAVRPNGMPESLIKYDARQKDSVVSSVSELVE...   \n",
       "6784           KU702192  MDVDVELDSDNSSSLGSYGDADAGFQFEAHNTTAPNLSFFGFDASV...   \n",
       "\n",
       "          Unnamed: 4      Unnamed: 5  \\\n",
       "0                NaN             NaN   \n",
       "1                NaN             NaN   \n",
       "2     GXNCRFLQ Motif  Protein Length   \n",
       "3           GKNCRILQ           850.0   \n",
       "4           GKTCHFLQ           855.0   \n",
       "...              ...             ...   \n",
       "6780        GKNCRFLQ          1083.0   \n",
       "6781        GRNCRFLQ          1008.0   \n",
       "6782        GRNCRFLQ           394.0   \n",
       "6783        GKNCRFLQ           465.0   \n",
       "6784        GRNCRFLQ           638.0   \n",
       "\n",
       "                                             Unnamed: 6  \\\n",
       "0                                                   NaN   \n",
       "1                                                   NaN   \n",
       "2                                      Domain Structure   \n",
       "3     [N_terminus : (0, 0) ---> GAF_2 : (9, 166) ---...   \n",
       "4     [N_terminus : (0, 0) ---> GAF_2 : (26, 166) --...   \n",
       "...                                                 ...   \n",
       "6780  [N_terminus : (0, 0) ---> TLOV : (286, 663) --...   \n",
       "6781  [N_terminus : (0, 0) ---> TLOV : (232, 597) --...   \n",
       "6782  [N_terminus : (0, 0) ---> TLOV : (49, 354) ---...   \n",
       "6783  [N_terminus : (0, 0) ---> LOV : (142, 247) ---...   \n",
       "6784  [N_terminus : (0, 0) ---> LOV : (55, 166) --->...   \n",
       "\n",
       "                                 Unnamed: 7        Unnamed: 8  \\\n",
       "0                                       NaN               NaN   \n",
       "1                                       NaN               NaN   \n",
       "2                        Functional Cluster  Primary Effector   \n",
       "3     ['EAL', 'GAF', 'GGDEF', 'LOV', 'PAS']             GGDEF   \n",
       "4     ['EAL', 'GAF', 'GGDEF', 'LOV', 'PAS']             GGDEF   \n",
       "...                                     ...               ...   \n",
       "6780                    ['Pkinase', 'TLOV']           Pkinase   \n",
       "6781                    ['Pkinase', 'TLOV']           Pkinase   \n",
       "6782                               ['TLOV']         Short_LOV   \n",
       "6783                     ['LOV', 'Pkinase']           Pkinase   \n",
       "6784              ['F-box', 'Kelch', 'LOV']             F-box   \n",
       "\n",
       "                          Unnamed: 9    Unnamed: 10  \\\n",
       "0                                NaN            NaN   \n",
       "1                                NaN            NaN   \n",
       "2     Primary Effector Gene Ontology  Linker Length   \n",
       "3      Cyclic Nucleotide Biosythesis           12.0   \n",
       "4      Cyclic Nucleotide Biosythesis           12.0   \n",
       "...                              ...            ...   \n",
       "6780                 Kinase Activity           83.0   \n",
       "6781                 Kinase Activity           83.0   \n",
       "6782                       Short LOV            0.0   \n",
       "6783                 Kinase Activity           83.0   \n",
       "6784             Protein interaction           56.0   \n",
       "\n",
       "                                    Unnamed: 11  \\\n",
       "0                                           NaN   \n",
       "1                                           NaN   \n",
       "2     Number of Predicted Transmembrane Helices   \n",
       "3                                           0.0   \n",
       "4                                           0.0   \n",
       "...                                         ...   \n",
       "6780                                        0.0   \n",
       "6781                                        0.0   \n",
       "6782                                        0.0   \n",
       "6783                                        0.0   \n",
       "6784                                        0.0   \n",
       "\n",
       "                        Unnamed: 12  Unnamed: 13     Unnamed: 14  \\\n",
       "0                               NaN          NaN             NaN   \n",
       "1                               NaN          NaN             NaN   \n",
       "2     Transmembrane Helix Topology       Kingdom          Phylum   \n",
       "3                              None     Bacteria  Proteobacteria   \n",
       "4                              None     Bacteria  Proteobacteria   \n",
       "...                             ...          ...             ...   \n",
       "6780                           None  Land Plants       Pinophyta   \n",
       "6781                           None  Land Plants  Lycopodiophyta   \n",
       "6782                           None  Land Plants  Lycopodiophyta   \n",
       "6783                           None  Land Plants  Lycopodiophyta   \n",
       "6784                           None  Land Plants  Lycopodiophyta   \n",
       "\n",
       "              Unnamed: 15      Unnamed: 16      Unnamed: 17    Unnamed: 18  \\\n",
       "0                     NaN              NaN              NaN            NaN   \n",
       "1                     NaN              NaN              NaN            NaN   \n",
       "2                   Class            Order           Family          Genus   \n",
       "3     Gammaproteobacteria  Pseudomonadales    Moraxellaceae  Acinetobacter   \n",
       "4     Gammaproteobacteria  Pseudomonadales    Moraxellaceae  Acinetobacter   \n",
       "...                   ...              ...              ...            ...   \n",
       "6780            Pinopsida          Pinales         Taxaceae          Taxus   \n",
       "6781          Isoetopsida   Selaginellales  Selaginellaceae    Selaginella   \n",
       "6782          Isoetopsida   Selaginellales  Selaginellaceae    Selaginella   \n",
       "6783          Isoetopsida   Selaginellales  Selaginellaceae    Selaginella   \n",
       "6784          Isoetopsida   Selaginellales  Selaginellaceae    Selaginella   \n",
       "\n",
       "       Unnamed: 19  \n",
       "0              NaN  \n",
       "1              NaN  \n",
       "2          Species  \n",
       "3        baumannii  \n",
       "4        baumannii  \n",
       "...            ...  \n",
       "6780     cuspidata  \n",
       "6781  lepidophylla  \n",
       "6782  lepidophylla  \n",
       "6783  lepidophylla  \n",
       "6784  lepidophylla  \n",
       "\n",
       "[6785 rows x 20 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(XLS_msa)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  2. Generate the sequences for MSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "msa_align = 'original_data/msa/msa_align.txt'\n",
    "df = pd.read_csv(XLS_msa_seq_csv)\n",
    "\n",
    "# # Find the fourth column element of the row where the second column has the value 'Q2NB89'\n",
    "# target_value = df[df.iloc[:, 1] == 'Q2NB98'].iloc[0, 3]\n",
    "\n",
    "# # Extract the fourth column of all rows starting from index 3\n",
    "# selected_column = df.iloc[3:, 3]\n",
    "\n",
    "# # Prepend the target value to the selected column data\n",
    "# result_series = pd.concat([pd.Series([target_value]), selected_column])\n",
    "\n",
    "# # Write the result to the text file\n",
    "# result_series.to_csv(msa_align, index=False, header=False)\n",
    "\n",
    "# Initialize a list to store formatted strings\n",
    "formatted_rows = []\n",
    "\n",
    "# Check if there's a row where the second column has the value 'Q2NB89' and add it first\n",
    "if 'Q2NB98' in df.iloc[:, 1].values:\n",
    "    target_row = df[df.iloc[:, 1] == 'Q2NB98'].iloc[0]\n",
    "    formatted_rows.append('>' + str(target_row.iloc[1]) + '\\n' + str(target_row.iloc[3]))\n",
    "\n",
    "# Format and add the remaining rows starting from index 3\n",
    "for _, row in df.iloc[3:].iterrows():\n",
    "    formatted_rows.append('>' + str(row.iloc[1]) + '\\n' + str(row.iloc[3]))\n",
    "\n",
    "# Join the formatted rows with a newline character\n",
    "formatted_string = '\\n'.join(formatted_rows)\n",
    "\n",
    "# Write the result to the text file\n",
    "with open(msa_align, 'w', encoding='utf-8') as file:\n",
    "    file.write(formatted_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ">Here we generate the two files that will be used for msa. We are only interested in the LOV domain of the WT protein (Q2NB98) - from the 16th amino acid (GADD...) to the 168th (...KTLS)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# File paths\n",
    "WT_path = 'original_data/msa/WT_align.txt'  # File for the first two lines\n",
    "sequences_path = 'original_data/msa/msa_comparisons.txt'  # File for the rest of the lines\n",
    "\n",
    "# Read the original file\n",
    "with open(msa_align, 'r', encoding='utf-8') as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "# Modify the second line to keep only the 16th to 151st amino acids (assuming 1-based indexing)\n",
    "if len(lines) > 1 and len(lines[1]) >= 151:\n",
    "    lines[1] = lines[1][16:167] + '\\n'  # Adjusting for 0-based indexing in Python\n",
    "\n",
    "# Write the first two lines to the first file\n",
    "with open(WT_path, 'w', encoding='utf-8') as file:\n",
    "    file.writelines(lines[:2])\n",
    "\n",
    "# Write the rest of the lines to the second file\n",
    "with open(sequences_path, 'w', encoding='utf-8') as file:\n",
    "    file.writelines(lines[2:])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  2. Convert MSA fasta file to adequate .a2m format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Look at EVMutation git (2. Computing pairwise models of sequences) the alignment requirements are the folowing:\n",
    "The first step of the EVmutation method is to compute a pairwise model of sequences from a family sequence alignment, e.g. the included example file (example/msa_comparison.a2m). This alignment has to be in aligned FASTA/A2M format and must fulfill the following requirements:\n",
    "* The target sequence may not contain any gaps.\n",
    "* Columns that should be excluded from model inference (e.g. too many gaps) must be represented by lowercase characters. Gaps in these positions must be represented by a dot (\".\") rather than a dash (\"-\").\n",
    "* The identifier of the target sequence has to be passed to plmc with the -f parameter (see below)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_alignment(fasta_file, start, stop, dots_to_dashes=True):\n",
    "    \"\"\"\n",
    "    Processes a protein multiple sequence alignment.\n",
    "\n",
    "    Args:\n",
    "    fasta_file (str): Path to the input FASTA file.\n",
    "    start (int): Start index of the region of interest (1-based, inclusive).\n",
    "    stop (int): Stop index of the region of interest (1-based, inclusive).\n",
    "    dots_to_dashes (bool): If True, convert all dashes to dots in the sequences.\n",
    "                           If False, remove positions with dashes in the WT sequence.\n",
    "\n",
    "    Raises:\n",
    "    ValueError: If start or stop indices are out of bounds.\n",
    "    FileNotFoundError: If the input file is not found.\n",
    "    \"\"\"\n",
    "\n",
    "    try:\n",
    "        # Read the alignment from the FASTA file\n",
    "        alignment = AlignIO.read(fasta_file, \"fasta\")\n",
    "    except FileNotFoundError:\n",
    "        raise FileNotFoundError(f\"File {fasta_file} not found.\")\n",
    "\n",
    "    # Get the WT sequence (first sequence in the alignment)\n",
    "    wt_sequence = str(alignment[0].seq)\n",
    "\n",
    "    # Check the dot-to-dash conversion or dash removal requirement\n",
    "    if dots_to_dashes:\n",
    "        wt_sequence = wt_sequence.replace('-', '.')\n",
    "        # Convert dashes to dots in all sequences\n",
    "        for record in alignment:\n",
    "            record.seq = Seq(str(record.seq).replace('-', '.'))\n",
    "    else:\n",
    "        # Remove positions where WT sequence has a dash\n",
    "        non_dash_positions = [i for i, char in enumerate(wt_sequence) if char != '-']\n",
    "        new_alignment = MultipleSeqAlignment([])\n",
    "        for record in alignment:\n",
    "            new_seq = ''.join(record.seq[i] for i in non_dash_positions)\n",
    "            new_alignment.append(SeqRecord(Seq(new_seq), id=record.id, description=record.description))\n",
    "        alignment = new_alignment\n",
    "        wt_sequence = ''.join(char for char in wt_sequence if char != '-')\n",
    "\n",
    "    # Find amino acid positions to identify start and stop\n",
    "    aa_positions = [i for i, char in enumerate(wt_sequence) if char != '.']\n",
    "\n",
    "    # Check if start and stop are within the bounds of the WT sequence\n",
    "    if start < 1 or stop > len(aa_positions):\n",
    "        raise ValueError(\"Start and stop indices are out of the sequence bounds.\")\n",
    "\n",
    "    # Convert start and stop from 1-based to 0-based and from amino acid count to alignment position\n",
    "    start_index = aa_positions[start - 1]\n",
    "    stop_index = aa_positions[stop - 1] if stop <= len(aa_positions) else len(wt_sequence) - 1\n",
    "\n",
    "    # Process each sequence in the alignment\n",
    "    processed_alignment = MultipleSeqAlignment([])\n",
    "    for record in alignment:\n",
    "        sequence = str(record.seq)\n",
    "        processed_seq = (\n",
    "            sequence[:start_index].lower() +\n",
    "            sequence[start_index:stop_index + 1] +\n",
    "            sequence[stop_index + 1:].lower()\n",
    "        )\n",
    "        processed_record = SeqRecord(Seq(processed_seq), id=record.id, description=record.description)\n",
    "        processed_alignment.append(processed_record)\n",
    "\n",
    "    # Save the processed alignment to an .a2m file\n",
    "    output_file = fasta_file.rsplit('.', 1)[0] + \".a2m\"\n",
    "    AlignIO.write(processed_alignment, output_file, \"fa[sta]\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ">Let's apply this function to the fasta file (LOV domains starting with EL222 aligned.fasta)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "File original_data\\msa\\LOV_domains_starting_with_EL222_aligned.fasta not found.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb Cell 15\u001b[0m line \u001b[0;36m1\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X20sZmlsZQ%3D%3D?line=16'>17</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X20sZmlsZQ%3D%3D?line=17'>18</a>\u001b[0m     \u001b[39m# Read the alignment from the FASTA file\u001b[39;00m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X20sZmlsZQ%3D%3D?line=18'>19</a>\u001b[0m     alignment \u001b[39m=\u001b[39m AlignIO\u001b[39m.\u001b[39;49mread(fasta_file, \u001b[39m\"\u001b[39;49m\u001b[39mfasta\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X20sZmlsZQ%3D%3D?line=19'>20</a>\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mFileNotFoundError\u001b[39;00m:\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/Bio/AlignIO/__init__.py:383\u001b[0m, in \u001b[0;36mread\u001b[0;34m(handle, format, seq_count)\u001b[0m\n\u001b[1;32m    382\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 383\u001b[0m     alignment \u001b[39m=\u001b[39m \u001b[39mnext\u001b[39;49m(iterator)\n\u001b[1;32m    384\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mStopIteration\u001b[39;00m:\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/Bio/AlignIO/__init__.py:322\u001b[0m, in \u001b[0;36mparse\u001b[0;34m(handle, format, seq_count)\u001b[0m\n\u001b[1;32m    320\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mTypeError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mNeed integer for seq_count (sequences per alignment)\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m--> 322\u001b[0m \u001b[39mwith\u001b[39;00m as_handle(handle) \u001b[39mas\u001b[39;00m fp:\n\u001b[1;32m    323\u001b[0m     \u001b[39m# Map the file format to a sequence iterator:\u001b[39;00m\n\u001b[1;32m    324\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mformat\u001b[39m \u001b[39min\u001b[39;00m _FormatToIterator:\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.8/contextlib.py:113\u001b[0m, in \u001b[0;36m_GeneratorContextManager.__enter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    112\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 113\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mnext\u001b[39;49m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mgen)\n\u001b[1;32m    114\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mStopIteration\u001b[39;00m:\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/Bio/File.py:72\u001b[0m, in \u001b[0;36mas_handle\u001b[0;34m(handleish, mode, **kwargs)\u001b[0m\n\u001b[1;32m     71\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m---> 72\u001b[0m     \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39;49m(handleish, mode, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs) \u001b[39mas\u001b[39;00m fp:\n\u001b[1;32m     73\u001b[0m         \u001b[39myield\u001b[39;00m fp\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'original_data\\\\msa\\\\LOV_domains_starting_with_EL222_aligned.fasta'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb Cell 15\u001b[0m line \u001b[0;36m4\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X20sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m start_index \u001b[39m=\u001b[39m \u001b[39m17\u001b[39m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X20sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m stop_index \u001b[39m=\u001b[39m \u001b[39m169\u001b[39m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X20sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m process_alignment(LOV_msa_fasta, start_index, stop_index, dots_to_dashes\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m)\n",
      "\u001b[1;32m/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb Cell 15\u001b[0m line \u001b[0;36m2\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X20sZmlsZQ%3D%3D?line=18'>19</a>\u001b[0m     alignment \u001b[39m=\u001b[39m AlignIO\u001b[39m.\u001b[39mread(fasta_file, \u001b[39m\"\u001b[39m\u001b[39mfasta\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X20sZmlsZQ%3D%3D?line=19'>20</a>\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mFileNotFoundError\u001b[39;00m:\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X20sZmlsZQ%3D%3D?line=20'>21</a>\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mFileNotFoundError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mFile \u001b[39m\u001b[39m{\u001b[39;00mfasta_file\u001b[39m}\u001b[39;00m\u001b[39m not found.\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X20sZmlsZQ%3D%3D?line=22'>23</a>\u001b[0m \u001b[39m# Get the WT sequence (first sequence in the alignment)\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X20sZmlsZQ%3D%3D?line=23'>24</a>\u001b[0m wt_sequence \u001b[39m=\u001b[39m \u001b[39mstr\u001b[39m(alignment[\u001b[39m0\u001b[39m]\u001b[39m.\u001b[39mseq)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: File original_data\\msa\\LOV_domains_starting_with_EL222_aligned.fasta not found."
     ]
    }
   ],
   "source": [
    "LOV_msa_fasta = 'original_data\\msa\\LOV_domains_starting_with_EL222_aligned.fasta'\n",
    "start_index = 17\n",
    "stop_index = 169\n",
    "process_alignment(LOV_msa_fasta, start_index, stop_index, dots_to_dashes=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Unknown format 'a2m'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb Cell 17\u001b[0m line \u001b[0;36m1\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X23sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m input_fasta \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39minput_data.fasta\u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X23sZmlsZQ%3D%3D?line=15'>16</a>\u001b[0m output_a2m \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39moutput.a2m\u001b[39m\u001b[39m'\u001b[39m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X23sZmlsZQ%3D%3D?line=17'>18</a>\u001b[0m fasta_to_a2m(input_fasta, output_a2m)\n",
      "\u001b[1;32m/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb Cell 17\u001b[0m line \u001b[0;36m1\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X23sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m alignment \u001b[39m=\u001b[39m MultipleSeqAlignment(records)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X23sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m \u001b[39m# Save the alignment in A2M format using the convert function\u001b[39;00m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X23sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m AlignIO\u001b[39m.\u001b[39;49mconvert(alignment, \u001b[39m\"\u001b[39;49m\u001b[39mfasta\u001b[39;49m\u001b[39m\"\u001b[39;49m, output_file, \u001b[39m\"\u001b[39;49m\u001b[39ma2m\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X23sZmlsZQ%3D%3D?line=11'>12</a>\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mConversion successful. A2M file saved as \u001b[39m\u001b[39m{\u001b[39;00moutput_file\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/Bio/AlignIO/__init__.py:471\u001b[0m, in \u001b[0;36mconvert\u001b[0;34m(in_file, in_format, out_file, out_format, molecule_type)\u001b[0m\n\u001b[1;32m    468\u001b[0m         \u001b[39mreturn\u001b[39;00m alignment\n\u001b[1;32m    470\u001b[0m     alignments \u001b[39m=\u001b[39m (over_ride(_) \u001b[39mfor\u001b[39;00m _ \u001b[39min\u001b[39;00m alignments)\n\u001b[0;32m--> 471\u001b[0m \u001b[39mreturn\u001b[39;00m write(alignments, out_file, out_format)\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/Bio/AlignIO/__init__.py:230\u001b[0m, in \u001b[0;36mwrite\u001b[0;34m(alignments, handle, format)\u001b[0m\n\u001b[1;32m    228\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mReading format \u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mformat\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m is supported, but not writing\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    229\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 230\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mUnknown format \u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mformat\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    232\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(count, \u001b[39mint\u001b[39m):\n\u001b[1;32m    233\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mRuntimeError\u001b[39;00m(\n\u001b[1;32m    234\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mInternal error - the underlying \u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    235\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mwriter should have returned the alignment count, not \u001b[39m\u001b[39m%r\u001b[39;00m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m (\u001b[39mformat\u001b[39m, count)\n\u001b[1;32m    236\u001b[0m     )\n",
      "\u001b[0;31mValueError\u001b[0m: Unknown format 'a2m'"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio import AlignIO\n",
    "\n",
    "# Function to convert a FASTA file to A2M format\n",
    "def fasta_to_a2m(input_file, output_file):\n",
    "    records = list(SeqIO.parse(input_file, \"fasta\"))\n",
    "    alignment = MultipleSeqAlignment(records)\n",
    "\n",
    "    # Save the alignment in A2M format using the convert function\n",
    "    AlignIO.convert(alignment, \"fasta\", output_file, \"a2m\")\n",
    "    print(f\"Conversion successful. A2M file saved as {output_file}\")\n",
    "\n",
    "# Replace 'input.fasta' and 'output.a2m' with your file names\n",
    "input_fasta = 'input_data.fasta'\n",
    "output_a2m = 'output.a2m'\n",
    "\n",
    "fasta_to_a2m(input_fasta, output_a2m)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO\n",
    "from Bio.Align.a2m import AlignmentWriter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the alignment from a FASTA file\n",
    "alignment = AlignIO.read(\"input_data.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Unknown format 'fas[ta]'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb Cell 21\u001b[0m line \u001b[0;36m3\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X26sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39m# Write the alignment to an A2M file\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X26sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39moutput_alignment.a2m\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mw\u001b[39m\u001b[39m\"\u001b[39m) \u001b[39mas\u001b[39;00m handle:\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/chloedaul/Desktop/ml-project-2-oknneighbors-main/msa_exploration.ipynb#X26sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m     AlignIO\u001b[39m.\u001b[39;49mwrite([alignment], handle, \u001b[39m\"\u001b[39;49m\u001b[39mfas[ta]\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/Bio/AlignIO/__init__.py:230\u001b[0m, in \u001b[0;36mwrite\u001b[0;34m(alignments, handle, format)\u001b[0m\n\u001b[1;32m    228\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mReading format \u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mformat\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m is supported, but not writing\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    229\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 230\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mUnknown format \u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mformat\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    232\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(count, \u001b[39mint\u001b[39m):\n\u001b[1;32m    233\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mRuntimeError\u001b[39;00m(\n\u001b[1;32m    234\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mInternal error - the underlying \u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    235\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mwriter should have returned the alignment count, not \u001b[39m\u001b[39m%r\u001b[39;00m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m (\u001b[39mformat\u001b[39m, count)\n\u001b[1;32m    236\u001b[0m     )\n",
      "\u001b[0;31mValueError\u001b[0m: Unknown format 'fas[ta]'"
     ]
    }
   ],
   "source": [
    "# Write the alignment to an A2M file\n",
    "with open(\"output_alignment.a2m\", \"w\") as handle:\n",
    "    AlignIO.write([alignment], handle, \"fas[ta]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ada",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
